pkg/pidhandle/pidhandle_linux.go

```diff
@@ -0,0 +1,186 @@
+//go:build linux
+
+// Package for handling processes and PIDs.
+package pidhandle
+
+import (
+	"encoding/hex"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+)
+
+type pidfdHandle struct {
+	pidfd        int
+	normalHandle pidHandle
+}
+
+// Store the "unix." methods in variables so we can mock them
+// in the unit-tests and test out different return value.
+var (
+	pidfdOpen       = unix.PidfdOpen
+	newFileHandle   = unix.NewFileHandle
+	openByHandleAt  = unix.OpenByHandleAt
+	nameToHandleAt  = unix.NameToHandleAt
+	pidfdSendSignal = unix.PidfdSendSignal
+)
+
+// Creates new PIDHandle for a given process pid.
+func NewPIDHandle(pid int) (PIDHandle, error) {
+	// Use the pidfd to obtain the file-descriptor pointing to the process.
+	pidData := ""
+	pidfd, err := pidfdOpen(pid, 0)
+	if err != nil {
+		switch err {
+		case unix.ENOSYS:
+			// Do not fail if PidFdOpen is not supported, we will
+			// fallback to process start-time later.
+
+		case unix.ESRCH:
+			// The process does not exist, so any future call of Kill
+			// or IsAlive should return unix.ESRCH, even if the pid is
+			// recycled in the future. Let's note it in the pidData.
+			pidData = noSuchProcessID
+
+		case unix.EINVAL:
+			// The PidfdOpen returns EINVAL if pid is invalid or if it refers
+			// to a thread and not to process. This is not a valid PID for
+			// PIDHandle and it most likely means the pid has been recycled
+			// (or there is a programming error). We therefore store
+			// noSuchProcessID into pidData to return unix.ESRCH in
+			// the future Kill or IsAlive calls.
+			pidData = noSuchProcessID
+
+		default:
+			return &pidHandle{}, err
```

Comment:
This should be `return nil,err` but also the error that is returned will just be the errno string which is hard to debug if these things ever fail.
It is best to wrap this with something like `fmt.Errorf("pidfd_open: %w", err)` then if such error is ever reported we at least know which syscall failed here without having to strace.